crates/cuda-common/src/memory_manager/vm_pool.rs

```diff
@@ -161,142 +137,340 @@ impl VirtualMemoryPool {
             }
         }
         // Potential merge with previous neighbor
-        if let Some((&prev_ptr, &prev_size)) = self.free_regions.range(..ptr).next_back() {
-            if prev_ptr + prev_size as u64 == ptr {
-                self.free_regions.remove(&prev_ptr);
+        if let Some((&prev_ptr, prev_region)) = self.free_regions.range(..ptr).next_back() {
+            if prev_ptr + prev_region.size as u64 == ptr && prev_region.stream_id == stream_id {
+                let prev_region = self.free_regions.remove(&prev_ptr).unwrap();
+                if prev_region.event.as_raw_handle() > event.as_raw_handle() {
+                    event = prev_region.event;
+                }
                 ptr = prev_ptr;
-                size += prev_size;
+                size += prev_region.size;
             }
         }
-        self.free_regions.insert(ptr, size);
+        self.free_regions.insert(
+            ptr,
+            FreeRegion {
+                size,
+                event,
+                stream_id,
+            },
+        );
     }
 
+    /// Defragments the pool by consolidating free regions at the end of virtual address space.
+    /// Remaps pages to create one large contiguous free region.
     fn defragment_or_create_new_pages(
         &mut self,
         requested: usize,
+        stream_id: CudaStreamId,
     ) -> Result<Option<CUdeviceptr>, MemoryError> {
-        if self.free_regions.is_empty() {
-            self.create_new_pages(requested)?;
-            return Ok(self.free_regions.iter().next_back().map(|(&p, _)| p));
+        if requested == 0 {
+            return Ok(None);
         }
 
-        let mut to_defrag: Vec<(CUdeviceptr, usize)> =
-            self.free_regions.iter().map(|(&p, &s)| (p, s)).collect();
-        let mut sum = 0;
-        // If last free region is at the virtual end, it will be used for defragmentation
-        if to_defrag
+        tracing::info!(
+            "Defragging or creating new pages: requested={}, stream_id={}",
+            requested,
+            stream_id
+        );
+
+        // 2.a. Defrag current stream (newest first)
+        let mut current_stream_to_defrag: Vec<(CUdeviceptr, usize, CudaEventHandle)> = self
+            .free_regions
+            .iter()
+            .filter(|(_, r)| r.stream_id == stream_id)
+            .map(|(addr, r)| (*addr, r.size, r.event.as_raw_handle()))
+            .collect();
+
+        // If last free region is at the virtual end, use it as base for defragmentation
+        let (mut defrag_start, mut accumulated_size) = if current_stream_to_defrag
             .last()
-            .is_some_and(|(p, s)| p + *s as u64 == self.curr_end)
+            .is_some_and(|(addr, size, _)| *addr + *size as u64 == self.curr_end)
         {
-            sum = to_defrag.pop().unwrap().1;
+            let (addr, size, _) = current_stream_to_defrag.pop().unwrap();
+            (addr, size)
+        } else {
+            (self.curr_end, 0)
```

Comment:
edge case: what if we go to 2.b, and actually at `self.curr_end` there is a free region on another stream that is **not** completed?

Reply:
nvm I think this should never happen because self.curr_end should really be the end of all VA